entity/contents.cc

```diff
@@ -457,4 +457,114 @@ bool ClipRestoreContents::Render(const ContentContext& renderer,
   return true;
 }
 
+/*******************************************************************************
+ ******* TextContents
+ ******************************************************************************/
+
+TextContents::TextContents() = default;
+
+TextContents::~TextContents() = default;
+
+void TextContents::SetTextFrame(TextFrame frame) {
+  frame_ = std::move(frame);
+}
+
+void TextContents::SetGlyphAtlas(std::shared_ptr<GlyphAtlas> atlas) {
+  atlas_ = std::move(atlas);
+}
+
+bool TextContents::Render(const ContentContext& renderer,
+                          const Entity& entity,
+                          RenderPass& pass) const {
+  if (!atlas_ || !atlas_->IsValid()) {
+    VALIDATION_LOG << "Cannot render glyphs without prepared atlas.";
+    return false;
+  }
+
+  using VS = GlyphAtlasPipeline::VertexShader;
+  using FS = GlyphAtlasPipeline::FragmentShader;
+
+  // Information shared by all glyph draw calls.
+  Command cmd;
+  cmd.label = "Glyph";
+  cmd.primitive_type = PrimitiveType::kTriangle;
+  cmd.pipeline = renderer.GetGlyphAtlasPipeline(OptionsFromPass(pass));
+  cmd.stencil_reference = entity.GetStencilDepth();
+
+  // Common vertex uniforms for all glyphs.
+  VS::FrameInfo frame_info;
+  frame_info.mvp = Matrix::MakeOrthographic(pass.GetRenderTargetSize()) *
+                   entity.GetTransformation();
+  frame_info.atlas_size =
+      Point{static_cast<Scalar>(atlas_->GetTexture()->GetSize().width),
+            static_cast<Scalar>(atlas_->GetTexture()->GetSize().height)};
+  VS::BindFrameInfo(cmd, pass.GetTransientsBuffer().EmplaceUniform(frame_info));
+
+  // Common fragment uniforms for all glyphs.
+  FS::BindGlyphAtlasSampler(
+      cmd,                                                        // command
+      atlas_->GetTexture(),                                       // texture
+      renderer.GetContext()->GetSamplerLibrary()->GetSampler({})  // sampler
+  );
+
+  // Common vertex information for all glyphs.
```

Comment:
It'd be helpful to me if this comment explained why we're addding a dummy 0x1 rect here

Reply:
The rect is not a dummy. Since we are drawing glyphs individually (to be batched later), we don't want to give each glyph unique vertex information. So we give all glyphs the same vertices but each gets its size in its uniform data. The vertex shader then transforms this unit sized box correctly. This also helps in the fragment shader where the interpolated vertex information (already unit-sized) is used to sample the texture from the atlas.

Reply:
Ahhh I see, we're basically giving it a reference point with this?

Reply:
I added a clarifying comment. The unit-sized box is doing triple duty. 1: Making it so that all glyphs get the same vertex data (and hence the data needs to be generated just once), 2: Actually specifying a quad so that the glyph can be drawn, and, 3: Helping out the fragment shader by not making it need additional coordinates to sample from the glyph atlas and making it easier (without having to apply an inverse scale) to just apply a scale-perspective transform for the glyph with an offset.